axum-extra/src/middleware/opentelemetry.rs

```diff
@@ -0,0 +1,513 @@
+///     // we must use `into_make_service_with_connect_info` for `opentelemetry_tracing_layer` to
+///     // access the client ip
+///     .serve(app.into_make_service_with_connect_info::<SocketAddr, _>())
+///     .await
+///     .expect("server failed");
+/// # };
+/// ```
+///
+/// # Complete example
+///
+/// See the "opentelemetry-jaeger" example for a complete setup that includes an OpenTelemetry
+/// pipeline sending traces to jaeger.
+///
+/// [otel]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md
+/// [`Router::into_make_service_with_connect_info`]: axum::Router::into_make_service_with_connect_info
+/// [`SetRequestIdLayer`]: tower_http::request_id::SetRequestIdLayer
+pub fn opentelemetry_tracing_layer() -> TraceLayer<
+    SharedClassifier<ServerErrorsAsFailures>,
+    OtelMakeSpan,
+    OtelOnRequest,
+    OtelOnResponse,
+    OtelOnBodyChunk,
+    OtelOnEos,
+    OtelOnFailure,
+> {
+    TraceLayer::new_for_http()
+        .make_span_with(OtelMakeSpan)
+        .on_request(OtelOnRequest)
+        .on_response(OtelOnResponse)
+        .on_body_chunk(OtelOnBodyChunk)
+        .on_eos(OtelOnEos)
+        .on_failure(OtelOnFailure)
+}
+
+/// A [`MakeSpan`] that creates tracing spans using [OpenTelemetry's conventional field names][otel].
+///
+/// [otel]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md
+#[derive(Clone, Copy, Debug)]
+pub struct OtelMakeSpan;
+
+impl<B> MakeSpan<B> for OtelMakeSpan {
+    fn make_span(&mut self, req: &Request<B>) -> Span {
+        let user_agent = req
+            .headers()
+            .get(header::USER_AGENT)
+            .map_or("", |h| h.to_str().unwrap_or(""));
+
+        let host = req
+            .headers()
+            .get(header::HOST)
+            .map_or("", |h| h.to_str().unwrap_or(""));
+
+        let scheme = req
+            .uri()
+            .scheme()
+            .map_or_else(|| "HTTP".into(), http_scheme);
+
+        let http_route = if let Some(matched_path) = req.extensions().get::<MatchedPath>() {
+            matched_path.as_str().to_owned()
+        } else if let Some(uri) = req.extensions().get::<OriginalUri>() {
+            uri.0.path().to_owned()
+        } else {
+            req.uri().path().to_owned()
+        };
+
+        let uri = if let Some(uri) = req.extensions().get::<OriginalUri>() {
+            uri.0.clone()
+        } else {
+            req.uri().clone()
+        };
+        let http_target = uri
+            .path_and_query()
+            .map(|path_and_query| path_and_query.to_string())
+            .unwrap_or_else(|| uri.path().to_owned());
+
+        let client_ip = req
+            .extensions()
+            .get::<ConnectInfo<SocketAddr>>()
+            .map(|ConnectInfo(client_ip)| Cow::from(client_ip.to_string()))
+            .unwrap_or_default();
+
+        let request_id = req
```

Comment:
Does request_id can be disabled with a feature?

Reply:
Yeah sure but want to elaborate on why you'd want this?

Reply:
Oh yeah thats a good point. That kinda makes me feel we shouldn't include request ids at all, since trace ids serve the same purpose 🤔

Reply:
I've removed request ids.